### ISSUE TITLE
Suggestion on enforcing PDF positivity

### ISSUE THREAD

**Issue by APJansen** [NNPDF/nnpdf#1752]
Not sure if this is the right place to discuss it, but I have a suggestion for the implementation of the positivity constraints of the PDFs themselves.

As I understand it:
1. What comes out of the PDF before any rotations is in the flavor basis, u, ubar, d, dbar, s, sbar, c, g.
2. These all have to be individually positive, except from the charm.
3. This is currently enforced using lagrange multipliers in the loss, using the same architecture as for the observables that have to be positive.

It seems to me that this can be done in a way that
1. Actually guarantees positivity rather than just penalizes negativity
2. Is simpler and faster
by simply applying an (e.g.) exponential activation function to all but the charm components, inside the PDF model before the `FlavourToEvolution` rotation is done. (Similarly to how a model that predicts class probabilities ends in a softmax layer, except without the normalization, and excluding one component)

My only concern with that would be that because one component wouldn't be scaled like this, it might lead to a separation of scales. But I think that the pdf output is of order 1 (both at initialization and physically, so I don't think it'd get far beyond that during training), so that shouldn't be an issue.

I may well be overlooking something, so I'm curious to hear your thoughts. (Also, let me know if there is a better place to discuss this)

**Comment by scarlehoff** [comment]
> What comes out of the PDF before any rotations is in the flavor basis, u, ubar, d, dbar, s, sbar, c, g.

No, what comes out of the PDF is in the evolution basis (and it could be any other basis) so even if we were to accept that all flavours need to be positive for beyond LO PDFs the output of the NN doesn't have to be.

When we know that the output of the PDF is positive for sure (for instance in Leading Order PDF) we indeed enforce positivity at the level of the NN (note how the Leading Order PDF runcard uses `square` as the last activation layer and fits in the flavours basis) https://github.com/NNPDF/nnpdf/blob/7d5b9f0a05c783883cd9e8f73929bf65279ddbef/n3fit/runcards/reproduce_nnpdf40/NNPDF40_lo_as_01180.yml#LL225C1-L225C1 

Enforcing positivity through observable has an extra nice (theory) feature which is the fact that we know that for sure cross sections have to be positive (because physical cross sections have to be positive) regardless of what one might think of the PDF themselves.

>  (Also, let me know if there is a better place to discuss this)

This would be better discussed either at the Friday meeting / mailing list / amsterdam-milan meetings since how (strong) to impose positivity is more a discussion about the physics that should involve the rest of the collaboration than about the code itself.
